### PR TITLE
PCP-6608: cluster-api-provider-maas wipes API server FQDN empty IP set persisted to MAAS DNS when CP machine is transiently powered off, hash-cached so no self-recovery

### PR DESCRIPTION
**Pull request by AmitSahastra**
###  Problem                                                                                                                                                                                                                                                                 
                                                            
  When a control-plane MaasMachine was briefly in Deployed (powered off) state, reconcileDNSAttachments() built an empty desired-IP set, PUT it to the MAAS DNSResource (wiping all A records), and persisted SHA-256("") as last-applied-dns-hash. Subsequent reconciles 
  saw no hash drift and returned early — DNS stayed empty with no self-recovery path. Workers lost connectivity to the API server FQDN, CNI never initialized, and the controller pod itself eventually went Unknown (chicken-and-egg).
                                                                                                                                                                                                                                                                          
  Confirmed on vmo-eng-2025 (VMO Eng PCG): DNS wiped at 16:52:22 UTC, made visible by an unrelated power outage at 18:02 UTC. The e3b0c44…b855 annotation (SHA-256 of empty string) was the forensic indicator.                                                           
                                                 
###  Changes                                                                                                                                                                                                                                                                 
                                         
  controllers/maascluster_controller.go

  - Empty-IP guard (primary fix): If runningIpAddresses is empty after filtering, return immediately without touching DNS or updating the hash annotation. The existing MAAS A records are preserved through the flap. When the CP recovers, the annotation hash will     
  differ from the new desired set and DNS is corrected on the next reconcile.
  - Improved log messages: Track runningCPCount separately from machines with eligible IPs. Log messages now distinguish two silent-failure modes: "no running CP machines" (power-off flap) vs "CP machines running but no IPs match preferred subnets"                  
  (misconfiguration).                                                                                                                                                                                                                                                     
  - MAAS drift detection: DNS resource is fetched before the hash check. When the annotation hash matches, MAAS state is still verified via IsDriftDetected(). If MAAS has drifted (e.g. records manually wiped while desired set was unchanged), the controller forces a
  re-sync instead of returning early forever.                                                                                                                                                                                                                             
                                                  
  pkg/maas/dns/dns.go                                                                                                                                                                                                                                                     
                                         
  - IsDriftDetected(): New method comparing a pre-fetched DNS resource's current IP set against a desired list. Used by the controller's drift check.                                                                                                                     
  - Defence-in-depth guard in updateResourceIPs(): Returns an error if the desired set is empty, so any future caller that bypasses the controller-level guard cannot accidentally wipe DNS.
                                                                                                                                                                                                                                                                          
###  Failure modes addressed                                                                                                                                                                                                                                                 
<img width="1323" height="267" alt="Screenshot 2026-05-08 at 11 22 27 AM" src="https://github.com/user-attachments/assets/f94c077e-85a4-44fa-b342-71e1611f4193" />

Remaining gap (not in this PR)                                                                                                                                                                                                                                          
                                                  
  If preferredSubnets is misconfigured such that all CP IPs are always filtered, DNS is preserved but no IP is ever written. The log message now surfaces this explicitly; a follow-up could add a condition/event on the MaasCluster object.

### Tests:

- [x] Empty-IP guard with existingCPCount / runningCPCount distinction
- [x] MAAS drift detection and recovery
- [x] IP deduplication before hashing


### Unit Tests:
•  TestIsRunning — All machine states including nil state
•  TestIsControlPlaneMachine — CP label detection
•  TestGetExternalMachineIP — Subnet filtering
•  TestReconcileDNSAttachments:
◦  CP powered off → DNS preserved
◦  No CP machines → DNS cleared
◦  CP with DeletionTimestamp → DNS cleared
◦  CP running but no ExternalIP → DNS preserved
◦  CP running, no prior annotation → DNS updated
◦  Hash match + MAAS in sync → skip update
◦  Hash match + MAAS drifted → force re-sync
◦  GetDNSResource error propagated
•  IsDriftDetected tests — matching IPs, empty MAAS, different IPs, extra IPs, duplicates, empty strings

### CODE REVIEW

File: pkg/maas/dns/dns.go [hunk 1]
```diff
@@ -83,6 +101,11 @@ func (s *Service) updateResourceIPs(dnsResource maasclient.DNSResource, IPs []st
 		}
 	}
 
+	// Refuse to wipe all A records; callers must guard before reaching here.
+	if desired.Len() == 0 {
+		return false, errors.New("refusing to PUT empty IP set to MAAS DNS resource")
+	}
+
```
[the remark this answers is not in the document]
Comment (AmitSahastra) [reply pkg/maas/dns/dns.go]: Fixed. Removed the guard from `updateResourceIPs` — the helper should remain a general-purpose apply-desired-state function and must support intentional clearing (e.g. deprovisioning). The protection against accidental DNS wipe belongs exclusively in `reconcileDNSAttachments` where the intent is known. The populated-to-empty test has been restored to its original semantics.

File: pkg/maas/dns/dns.go [hunk 2]
```diff
@@ -59,6 +59,24 @@ func (s *Service) ReconcileDNS() error {
...
+		}
+	}
+	current := sets.NewString()
+	for _, addr := range dnsResource.IPAddresses() {
+		if a := addr.IP().String(); a != "" {
+			current.Insert(a)
+		}
+	}
+	return !desired.Equal(current)
+}
```
[the remark this answers is not in the document]
Comment (AmitSahastra) [reply pkg/maas/dns/dns.go:78]: Addressed. Four tests for `IsDriftDetected` were already added in this PR (equal sets, empty MAAS, different IPs, extra IPs). Based on this comment, two additional edge cases have been added: duplicate IPs in desired (deduplication via set means no spurious drift) and empty strings in desired (filtered out, no false drift signal). 15 tests now pass in the dns package.

File: controllers/maascluster_controller.go [hunk 3]
```diff
@@ -171,34 +171,56 @@ func (r *MaasClusterReconciler) reconcileDNSAttachments(clusterScope *scope.Clus
...
 	dnsResource, err := dnssvc.GetDNSResource()
 	if err != nil {
 		return errors.Wrap(err, "Unable to get the dns resource")
 	}
 
+	// Early-exit only when the annotation hash matches AND MAAS state agrees with desired IPs.
+	// Verifying MAAS state catches external drift (e.g. DNS records manually wiped while the
+	// desired set was unchanged, so the hash never changed).
+	desiredHash := infrautil.StableHashStringSlice(runningIpAddresses)
+	if clusterScope.MaasCluster.Annotations != nil && clusterScope.MaasCluster.Annotations[lastAppliedAnn] == desiredHash {
```
[the remark this answers is not in the document]
Comment (AmitSahastra) [reply controllers/maascluster_controller.go:243]: Fixed. Added `dedupStringSlice` and applied it to `runningIpAddresses` before computing the hash. The annotation now always represents the applied IP set, consistent with the set semantics inside `updateResourceIPs`. Without this, a transient duplicate IP (two CP machines reporting the same address) would produce a hash mismatch on the next reconcile while MAAS state was unchanged, causing a spurious re-sync and unnecessary MAAS GET.

**Review by cPu1 — APPROVED** [review]

**Review by snehala27 — APPROVED** [review]
